.github/workflows/haskell.yml

```diff
@@ -80,28 +80,42 @@ jobs:
     - name: Cabal update
       run: cabal update
 
-    - name: Cabal configure
-      run: cabal configure --enable-tests --enable-benchmarks --write-ghc-environment-files=always
-
-    - uses: actions/cache@v4
-      if: matrix.os != 'macos-latest'
-      name: Cache cabal store
+    - name: Restore cache
+      uses: actions/cache/restore@v4
+      id: restore-cabal-cache
+      env:
+        cache-name: cache-cabal-build
       with:
         path: |
           ${{ steps.install-haskell.outputs.cabal-store }}
-        # cache is invalidated upon a change to cabal.project (and/or cabal.project.local), a bump to
-        # CABAL_CACHE_VERSION or after a week of inactivity
-        key: cache-${{ runner.os }}-${{ matrix.ghc }}-${{ env.CABAL_CACHE_VERSION }}-${{ hashFiles('cabal.project*') }}
-        # Restoring attempts are from current branch then master
+        # cache is invalidated upon a change to a cabal file, cabal.project (and/or cabal.project.local), or a bump to
+        # CABAL_CACHE_VERSION.
+        key: ${{ env.cache-name }}-${{ runner.os }}-${{ matrix.ghc }}-${{ env.CABAL_CACHE_VERSION }}-${{ hashFiles('**/*.cabal') }}-${{ hashFiles('cabal.project*') }}
```

Comment:
A more robust way would be to actually look at the cabal plan.

We don't want to invalidate cache on a change to any of the cabal files. For example bumping up a version in a cabal file should not cause us to rebuild all of the dependencies. We update index-state quite rarely, while touching cabal files we do all the time which means the cache will be almost always invalid.

FYI we don't do this dumb thing of having over-restrictive upper bounds on all dependencies, so version of dependencies is usually only affected by the `index-state`, not the cabal files.

I think we should steal the approach taken in cardano-base: 
https://github.com/IntersectMBO/cardano-base/blob/9b633d49ce5390591897576c97301319b936f6ed/.github/workflows/haskell.yml#L84-L108

```
$ cabal build all --enable-tests --dry-run --minimize-conflict-set
$ cat dist-newstyle/cache/plan.json | jq -r '."install-plan"[].id' | sort | uniq > dependencies.txt
```

Reply:
With the current changes a cache isn't invalidated when the cabal files change: it will restore an old cache, build any changed/new dependencies on top, and then store them under a new hash. However, using the approach you suggested above we would store new caches less often, so it makes sense to include

Reply:
> it will restore an old cache, build any changed/new dependencies on top, and then store them under a new hash

Yes, that's exactly what I meant by "invalidated".
Thank you for applying this suggestion.